README.md
# README 
## TreeMap 0.1


[![Elixir CI](https://github.com/iboard/tree_map/actions/workflows/elixir.yml/badge.svg)](https://github.com/iboard/tree_map/actions/workflows/elixir.yml)
[![Package version](http://img.shields.io/hexpm/v/tree_map.svg?style=flat)](https://hex.pm/packages/tree_map)
[![Package documentation](http://img.shields.io/badge/hex.pm-docs-green.svg?style=flat)](https://hexdocs.pm/tree_map)

A module to maintain a tree of the structure `TreeMap.Node` and to start root-nodes as 
supervised children of `TreeMap.RootServer`.

### Node Structure

```elixir
  %Node{ key: ..., value: ..., children: [%Node{},...]}
```

### Function to traverse `%Node{}`

```elixir
 root =
   TreeMap.new("1", "Root")
   |> TreeMap.add_child("1.1", "Sub One", [TreeMap.new("1.1.1", "SomeSubSub")])
   |> TreeMap.add_child("1.2", "Sub Two", [TreeMap.new("1.2.1", "Some other SubSub")])

 expected = ["1", "1.1", "1.1.1", "1.2", "1.2.1"]

 received =
   TreeMap.transduce(root, [], fn a, x -> a ++ [x] end, fn n, acc -> [n | acc] end)
   |> Enum.map(& &1.key)

 assert expected == received
```

### RootServer

`TreeMap.RootServer` is a `DynamicSupervisor` that can be used to maintain a tree of `TreeMap` nodes.

    iex> TreeMap.start_root_node("A", "Root")
    {:ok, pid}

    iex> TreeMap.list_roots()
    [%Node{}, ...]

    iex> TreeMap.find("key")
    %Node{}

    iex> TreeMap.drop_all()
    [:ok, ...]


## Installation

[Available in Hex](https://hex.pm/packages/tree_map), the package can be 
installed by adding `tree_map` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:tree_map, "~> 0.1.0"}
  ]
end
```




CHANGELOG.md
# Changelog

## 2023-09-15 Initial version 0.1.0



test/tree_map_test.exs
defmodule TreeMapTest do
  use ExUnit.Case
  doctest TreeMap
  alias TreeMap.Node

  test "initialize a TreeMap" do
    assert TreeMap.new() == %Node{key: nil, value: nil, children: []}
    assert TreeMap.new(1, :foo) == %Node{key: 1, value: :foo, children: []}
  end

  test "add child" do
    child = TreeMap.new("1.3", "added as a map")

    root =
      TreeMap.new("1", "Root")
      |> TreeMap.add_child("1.1", "Subtree A")
      |> TreeMap.add_child("1.2", "Subtree B")
      |> TreeMap.add_child(child)

    assert TreeMap.children(root) |> Enum.count() == 3
  end

  test "traverse" do
    root =
      TreeMap.new("1", "Root")
      |> TreeMap.add_child("1.1", "Subtree A")
      |> TreeMap.add_child("1.2", "Subtree B")

    assert TreeMap.traverse(root, fn node ->
             "#{node.key} #{node.value}"
           end) == [
             "1 Root",
             [
               "1.1 Subtree A",
               [],
               "1.2 Subtree B",
               []
             ]
           ]

    assert [2 | _] = TreeMap.traverse(root, fn node -> Enum.count(node.children) end)
  end

  test "flatten the tree with transduce" do
    root =
      TreeMap.new("1", "Root")
      |> TreeMap.add_child("1.1", "Sub One", [TreeMap.new("1.1.1", "SomeSubSub")])
      |> TreeMap.add_child("1.2", "Sub Two", [TreeMap.new("1.2.1", "Some other SubSub")])

    expected = ["1", "1.1", "1.1.1", "1.2", "1.2.1"]

    received =
      TreeMap.transduce(root, [], fn a, x -> a ++ [x] end, fn n, acc -> [n | acc] end)
      |> Enum.map(& &1.key)

    assert expected == received
  end

  test "deep sum with transduce" do
    root =
      TreeMap.new("1", "Root")
      |> TreeMap.add_child("1.1", "Sub One", [TreeMap.new("1.1.1", "SomeSubSub")])
      |> TreeMap.add_child("1.2", "Sub Two", [TreeMap.new("1.2.1", "Some other SubSub")])

    result =
      TreeMap.transduce(root, 38, fn a, x -> a + Enum.count(x.children) end, fn x, acc ->
        acc + x
      end)

    assert 42 == result
  end

  test "family tree" do
    prokop = create_family_tree()

    assert TreeMap.traverse(prokop, fn node ->
             "#{node.value}"
           end) == [
             "Prokop",
             [
               "Gerda",
               [
                 "Margit",
                 [],
                 "Heidi",
                 ["Alex", [], "Julian", []],
                 "Elke",
                 ["Hanna", [], "Jan", []]
               ],
               "Helmut",
               []
             ]
           ]
  end

  test "find" do
    prokop = create_family_tree()

    jan = TreeMap.find(prokop, "1.1.1.1")
    false = TreeMap.find(prokop, "not found")

    assert jan.value == "Jan"
  end

  defp create_family_tree() do
    helmut = TreeMap.new("1.2", "Helmut")

    margit = TreeMap.new("1.1.3", "Margit")

    jan = TreeMap.new("1.1.1.1", "Jan")
    hanna = TreeMap.new("1.1.1.2", "Hanna")
    elke = TreeMap.new("1.1.1", "Elke", [jan, hanna])

    julian = TreeMap.new("1.1.2.1", "Julian")
    alex = TreeMap.new("1.1.2.2", "Alex")
    heidi = TreeMap.new("1.1.2", "Heidi", [julian, alex])

    gerda = TreeMap.new("1.1", "Gerda", [elke, heidi, margit])

    TreeMap.new("1", "Prokop", [helmut, gerda])
  end
end


test/tree_map_server_test.exs
defmodule TreeMapServerTest do
  use ExUnit.Case

  alias TreeMap.Node

  describe "OTP basics" do
    setup %{} do
      TreeMap.drop_all()
      :ok
    end

    test "initialize the server at startup" do
      assert Process.whereis(TreeMap.Supervisor) != nil
    end

    test "add root nodes" do
      {:ok, prokop} = TreeMap.start_root_node("A", "Prokop")
      {:ok, altendorfer} = TreeMap.start_root_node("B", "Altendorfer")

      assert is_pid(prokop)
      assert is_pid(altendorfer)

      assert "Prokop" == TreeMap.value(prokop)
      assert "Altendorfer" == TreeMap.value(altendorfer)
    end

    test "list root servers" do
      prokop = create_family_tree()
      {:ok, prokop} = TreeMap.start_root_node(prokop)
      assert is_pid(prokop)
    end
  end

  describe "TreeMap with root nodes" do
    setup %{} do
      TreeMap.drop_all()
      {:ok, prokop} = TreeMap.start_root_node(create_family_tree("Prokop"))
      {:ok, altendorfer} = TreeMap.start_root_node(create_family_tree("Altendorfer"))
      {:ok, %{prokop: prokop, altendorfer: altendorfer}}
    end

    test "list nodes" do
      prokop = create_family_tree("Prokop")
      altendorfer = create_family_tree("Altendorfer")
      assert [^prokop, ^altendorfer] = TreeMap.list_roots()
    end

    test "find in all nodes" do
      # given two families
      prokop = create_family_tree("Prokop")
      altendorfer = create_family_tree("Altendorfer")

      # when search in a particular family
      jan = TreeMap.find(prokop, "1.1.1.1")
      andi = TreeMap.find(altendorfer, "2.2.2")

      # when search in all root_nodes
      ^jan = TreeMap.find("1.1.1.1")
      ^andi = TreeMap.find("2.2.2")
    end
  end

  describe "All nodes are processes" do
    setup %{} do
      TreeMap.drop_all()

      {:ok, project_one} = TreeMap.start_root_node("1", "Project One")
      {:ok, subproject_one_one} = TreeMap.start_root_node("1.1", "Subproject One One")
      {:ok, subproject_one_two} = TreeMap.start_root_node("1.2", "Subproject One Two")
      {:ok, project_two} = TreeMap.start_root_node("2", "Project Two")
      {:ok, subproject_two_one} = TreeMap.start_root_node("2.1", "Subproject Two One")

      TreeMap.add_child(project_one, subproject_one_one)
      TreeMap.add_child(project_one, subproject_one_two)
      TreeMap.add_child(project_two, subproject_two_one)

      :ok
    end

    test "Find node" do
      node = TreeMap.find("1.1")
      assert node.value == "Subproject One One"
    end

    test "Find roots only" do
      assert ["1", "2"] ==
               TreeMap.list_roots()
               |> Enum.map(&Node.key/1)
    end

    test "list all nodes with servers" do
      assert ["1", "1.1", "1.2", "2", "2.1"] ==
               TreeMap.list_all()
               |> Enum.map(&Node.key/1)
    end
  end

  defp create_family_tree(name \\ "Prokop")

  defp create_family_tree("Altendorfer") do
    grete =
      TreeMap.new("2.2", "Grete")
      |> TreeMap.add_child(TreeMap.new("2.2.1", "Rudi"))
      |> TreeMap.add_child(TreeMap.new("2.2.2", "Andreas"))

    TreeMap.new("2", "Altendorfer")
    |> TreeMap.add_child(TreeMap.new("2.1", "Rudolf"))
    |> TreeMap.add_child(grete)
  end

  defp create_family_tree("Prokop") do
    helmut = TreeMap.new("1.2", "Helmut")

    margit = TreeMap.new("1.1.3", "Margit")

    jan = TreeMap.new("1.1.1.1", "Jan")
    hanna = TreeMap.new("1.1.1.2", "Hanna")
    elke = TreeMap.new("1.1.1", "Elke") |> TreeMap.add_child(jan) |> TreeMap.add_child(hanna)

    julian = TreeMap.new("1.1.2.1", "Julian")
    alex = TreeMap.new("1.1.2.2", "Alex")
    heidi = TreeMap.new("1.1.2", "Heidi") |> TreeMap.add_child(julian) |> TreeMap.add_child(alex)

    gerda =
      TreeMap.new("1.1", "Gerda")
      |> TreeMap.add_child(elke)
      |> TreeMap.add_child(heidi)
      |> TreeMap.add_child(margit)

    TreeMap.new("1", "Prokop") |> TreeMap.add_child(helmut) |> TreeMap.add_child(gerda)
  end
end


mix.exs
defmodule TreeMap.MixProject do
  use Mix.Project

  def project do
    [
      app: :tree_map,
      version: "0.1.2",
      elixir: "~> 1.15",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      description: description(),
      package: package(),
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      docs: [
        main: "readme",
        homepage_url: "https://github.com/iboard/tree_map",
        source_url: "https://github.com/iboard/tree_map",
        extras: ["README.md", "LICENSE.md", "CHANGELOG.md"]
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger, :runtime_tools, :wx, :observer],
      mod: {TreeMap.Application, []}
    ]
  end

  defp description() do
    "Manage a tree of key, value, children struct"
  end

  defp package() do
    [
      files: ~w(lib test .formatter.exs mix.exs README* LICENSE*
                CHANGELOG*),
      licenses: ["Apache-2.0"],
      links: %{"GitHub" => "https://github.com/iboard/tree_map"}
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false}
    ]
  end
end


lib/tree_map/node.ex
defmodule TreeMap.Node do
  @moduledoc """
  `Node` can be used as a simple structure, representing a node in a tree
  or as a GenServer to store the Node structure.

  Node GenServer are supervised by `TreeMap.RootServer` and should 
  be started and stopped throug `TreeMap` module.
  """
  defstruct key: nil, value: nil, parent: nil, children: []
  alias __MODULE__

  use GenServer

  def start_link(%__MODULE__{} = node, opts \\ []) do
    GenServer.start_link(__MODULE__, node, opts)
  end

  @impl true
  def init(node) do
    {:ok, node}
  end

  @doc """
  Return the `%Node{}` struct of this process.
  """
  def get_node(pid) when is_pid(pid), do: GenServer.call(pid, :node)

  def get_node(%Node{} = node), do: node

  @doc """
  Returns the key of the `%Node{}` struct of this process.
  """
  def key(pid) when is_pid(pid), do: GenServer.call(pid, :key)

  def key(%Node{key: k}), do: k

  @doc """
  Returns the value of the `%Node{}` struct of this process.
  """
  def value(pid) when is_pid(pid), do: GenServer.call(pid, :value)

  def value(%Node{value: v}), do: v

  @doc """
  Returns the parent of the `%Node{}` struct of this process.
  """
  def parent(pid) when is_pid(pid), do: GenServer.call(pid, :parent)

  def parent(%Node{parent: p}), do: p

  @doc """
  Update the node of the given process.
  """
  def update(pid, node) when is_pid(pid) do
    GenServer.cast(pid, {:update, node})
    get_node(pid)
  end

  @impl true
  def handle_call(:node, _, %__MODULE__{} = state), do: {:reply, state, state}

  def handle_call(:key, _, %__MODULE__{key: key} = state), do: {:reply, key, state}

  def handle_call(:value, _, %__MODULE__{value: value} = state), do: {:reply, value, state}

  def handle_call(:parent, _, %__MODULE__{parent: parent} = state), do: {:reply, parent, state}

  @impl true
  def handle_cast({:update, node}, _state), do: {:noreply, node}
end


lib/tree_map/root_server.ex
defmodule TreeMap.RootServer do
  @moduledoc """
  A server to maintain a list of root nodes as a DynamicSupervisor.


  Don't call this module directly. Use the `TreeMap` module instead.

  ### Examples

      iex> TreeMap.list_roots()
      [%Node{}, ...]

      iex> TreeMap.find("key")
      %Node{}

      iex> TreeMap.drop_all()
      [:ok, ...]
  """
  use DynamicSupervisor
  alias TreeMap.Node

  @doc """
  Start the DynamicSupervisor as a singleton.
  """
  def start_link(init_arg) do
    DynamicSupervisor.start_link(__MODULE__, init_arg, name: __MODULE__)
  end

  @impl true
  def init(_init_arg) do
    DynamicSupervisor.init(strategy: :one_for_one)
  end

  @doc """
  Start a new root node as supervised child.
  """
  def start_child(node) do
    DynamicSupervisor.start_child(__MODULE__, {Node, node})
  end

  @doc """
  Return the root node of each child which has no parent (roots).
  """
  def list_roots() do
    for {:undefined, pid, :worker, [TreeMap.Node]} <- Supervisor.which_children(__MODULE__) do
      Node.get_node(pid)
    end
    |> Enum.filter(fn n -> n.parent == nil end)
  end

  @doc """
  Return the root node of each child.
  """
  def list_all() do
    for {:undefined, pid, :worker, [TreeMap.Node]} <- Supervisor.which_children(__MODULE__) do
      Node.get_node(pid)
    end
  end

  @doc """
  Stop all children.
  """
  def drop_all() do
    for {:undefined, pid, :worker, [TreeMap.Node]} <- Supervisor.which_children(__MODULE__) do
      DynamicSupervisor.terminate_child(__MODULE__, pid)
    end
  end
end


lib/tree_map.ex
defmodule TreeMap do
  @moduledoc """
  A module to maintain a tree of the structure

  ```elixir
  %Node{ key: ..., value: ..., children: [%TreeMap{},...]}
  ```

  and functions to traverse such a tree.

  ### Examples

  ```elixir
  root =
   TreeMap.new("1", "Root")
   |> TreeMap.add_child("1.1", "Sub One", [TreeMap.new("1.1.1", "SomeSubSub")])
   |> TreeMap.add_child("1.2", "Sub Two", [TreeMap.new("1.2.1", "Some other SubSub")])

  expected = ["1", "1.1", "1.1.1", "1.2", "1.2.1"]

  received =
   TreeMap.transduce(root, [], fn a, x -> a ++ [x] end, fn n, acc -> [n | acc] end)
   |> Enum.map(& &1.key)

  assert expected == received
  ```
  """

  alias TreeMap.{Node, RootServer}

  @doc """
  ### Examples

      iex> TreeMap.new()
      %TreeMap.Node{}

      iex> TreeMap.new(123, "hundretandtwentythree")
      %TreeMap.Node{ key: 123, value: "hundretandtwentythree"}

  """
  def new(key \\ nil, value \\ nil, children \\ []) do
    %Node{key: key, value: value, children: children}
  end

  @doc """
  Return the children of this node.
  """
  def children(tree_map), do: tree_map.children

  @doc """
  Prepend a new child to the children of this node.

  ### Example:
      iex> TreeMap.new("1", "Root") |> TreeMap.add_child("1.1", "Sub One")
      %TreeMap.Node{key: "1", value: "Root", children: [%TreeMap.Node{key: "1.1", value: "Sub One", parent: "1", children: []}]}
  """
  def add_child(tree_map, key, value, children \\ []) do
    %{
      tree_map
      | children: [%{new(key, value, children) | parent: Node.key(tree_map)} | tree_map.children]
    }
  end

  @doc """
  Prepend an exisitng child to the children of this node.

  ### Example:
      iex> child = TreeMap.new("1.1", "I exist")
      iex> TreeMap.new("1", "Root") |> TreeMap.add_child(child)
      %TreeMap.Node{key: "1", value: "Root", parent: nil, children: [%TreeMap.Node{key: "1.1", parent: "1", value: "I exist", children: []}]}
  """
  def add_child(tree_map, child_pid) when is_pid(tree_map) and is_pid(child_pid) do
    node = Node.get_node(tree_map)
    child = %{Node.get_node(child_pid) | parent: Node.key(tree_map)}
    node = add_child(node, child)
    Node.update(child_pid, child)
    Node.update(tree_map, node)
  end

  def add_child(tree_map, child_pid) when is_map(tree_map) and is_pid(child_pid) do
    child = %{Node.get_node(child_pid) | parent: Node.key(tree_map)}
    Node.update(child_pid, child)
    add_child(tree_map, child)
  end

  def add_child(tree_map, child) when is_pid(tree_map) and is_map(child) do
    node = Node.get_node(tree_map) |> add_child(child)
    Node.update(tree_map, node)
  end

  def add_child(tree_map, child) when is_map(tree_map) and is_map(child) do
    add_child(tree_map, child.key, child.value, child.children)
  end

  @doc """
  Traverse the tree map and apply the given function to each node.

  ### Example: 
      iex> root = TreeMap.new("1", "Root") |> TreeMap.add_child("1.1", "Sub One") |> TreeMap.add_child("1.2", "Sub Two")
      ...> TreeMap.traverse(root, fn x -> x.key end)
      ["1", ["1.1", [], "1.2", []]]
        
  """
  def traverse(tree_map, fun) do
    [
      fun.(tree_map)
      | [
          Enum.flat_map(Enum.reverse(tree_map.children), fn child ->
            traverse(child, fun)
          end)
        ]
    ]
  end

  @doc """
  Traverse the tree map and apply the given function to each node 
    and a reducer-function to it's children.

  ### Example: 
      iex> root = TreeMap.new("1", "Root") 
      ...>        |> TreeMap.add_child("1.1", "Sub One", [TreeMap.new("1.1.1", "SomeSubSub")]) 
      ...>        |> TreeMap.add_child("1.2", "Sub Two", [TreeMap.new("1.2.1", "Some other SubSub")])
      ...> TreeMap.transduce(root, 38, fn a, x -> a + Enum.count(x.children) end, fn x, acc -> acc + x end)
      42

      iex> root = TreeMap.new("1", "Root") 
      ...>        |> TreeMap.add_child("1.1", "Sub One", [TreeMap.new("1.1.1", "SomeSubSub")]) 
      ...>        |> TreeMap.add_child("1.2", "Sub Two", [TreeMap.new("1.2.1", "Some other SubSub")])
      ...> TreeMap.transduce(root, [], fn a, x -> a ++ [x.value] end, fn n, acc -> [n | acc] end) 
      ["Root", "Sub One", "SomeSubSub", "Sub Two", "Some other SubSub"]

      iex> root = TreeMap.new("1", "Root") 
      ...>        |> TreeMap.add_child("1.1", "Sub One", [TreeMap.new("1.1.1", "SomeSubSub")]) 
      ...>        |> TreeMap.add_child("1.2", "Sub Two", [TreeMap.new("1.2.1", "Some other SubSub")])
      ...> TreeMap.transduce(root, [], fn a, x -> a ++ [x] end, fn n, acc -> [n | acc] end) 
      ...> |> Enum.map(& &1.key)
      ["1", "1.1", "1.1.1", "1.2", "1.2.1"]
  """
  def transduce(tree_map, acc, modifier, reducer) when is_pid(tree_map) do
    Node.get_node(tree_map) |> transduce(acc, modifier, reducer)
  end

  def transduce(tree_map, acc, modifier, reducer) do
    Enum.reduce(Enum.reverse(tree_map.children), modifier.(acc, tree_map), fn n, a ->
      transduce(n, a, modifier, reducer)
    end)
  end

  @doc """
  Find a node by it's key or returns `false` in a given tree.
  """
  def find(tree_map, key) do
    transduce(tree_map, nil, fn a, n -> a || (n.key == key && n) end, fn n, acc ->
      (n.key == key && n) || acc
    end)
  end

  @doc """
  Find a node by it's key or returns `false` in all root nodes.
  """
  def find(key) do
    list_roots()
    |> Enum.reduce(false, fn node, found ->
      found || find(node, key)
    end)
  end

  @doc """
  Create a Node and start a RootServer for it.
  """
  def start_root_node(key, value) do
    node = new(key, value)
    RootServer.start_child(node)
  end

  @doc """
  Start a root_node for the given `%Node{}`.
  """
  def start_root_node(node) do
    RootServer.start_child(node)
  end

  @doc """
  Retreive the key of a given root node.
  """
  def key(root_pid) do
    Node.key(root_pid)
  end

  @doc """
  Retreive the value of a given root node.
  """
  def value(root_pid) do
    Node.value(root_pid)
  end

  @doc """
  Retreive the parent of a given root node.
  """
  def parent(root_pid) do
    Node.parent(root_pid)
  end

  @doc """
  List all root nodes from all RootServers.
  """
  def list_roots() do
    RootServer.list_roots()
  end

  @doc """
  List all nodes from all RootServers.
  """
  def list_all() do
    RootServer.list_all()
  end

  @doc """
  Kill all RootServers.
  """
  def drop_all() do
    RootServer.drop_all()
  end
end
